twoliter/embedded/img2img

```diff
@@ -1,3 +1,318 @@
+mapfile -t new_root_artifacts <<<"$(find "${ROOT_MOUNT}" -type f)"
+
+# TODO - Error if no artifacts since there will always at least the CA bundle.
+if [[ -z "${new_root_artifacts[0]}" ]]; then
+  echo "Nothing to write, skipping..."
+else
+  # Write files from the root mount to the root image.
+  for artifact in "${new_root_artifacts[@]}"; do
+    cat <<EOF | debugfs -w -f - "${ROOT_IMAGE}"
+rm ${artifact#"${ROOT_MOUNT}"}
+write ${artifact} ${artifact#"${ROOT_MOUNT}"}
+ea_set ${artifact#"${ROOT_MOUNT}"} security.selinux system_u:object_r:os_t:s0
+EOF
+  done
+fi
+
+# Validate and write root image back to the OS image.
+check_image_size "${ROOT_IMAGE}" "${partsize["ROOT-A"]}"
+dd if="${ROOT_IMAGE}" of="${OS_IMAGE}" \
+  conv=notrunc bs=1M seek="${partoff["ROOT-A"]}"
+
+# Generate a new root verity.
+declare -a DM_VERITY_ROOT
+generate_verity_root "${ROOT_IMAGE}" "${VERITY_IMAGE}" \
+  "${partsize["HASH-A"]}" DM_VERITY_ROOT
+
+# Validate and write root verity back to the OS image. The image size check
+# isn't needed here as it's already done in `generate_verity_root`.
+dd if="${VERITY_IMAGE}" of="${OS_IMAGE}" \
+  conv=notrunc bs=1M seek="${partoff["HASH-A"]}"
+
+###############################################################################
+# Section 5: maybe secure boot
+
+if [[ "${UEFI_SECURE_BOOT}" == "yes" ]]; then
+  pushd "${EFI_MOUNT}/EFI/BOOT" >/dev/null
+  shims=(boot*.efi)
+  shim="${shims[0]}"
+  grubs=(grub*.efi)
+  grub="${grubs[0]}"
+  mokms=(mm*.efi)
+  mokm="${mokms[0]}"
+
+  # Do the setup required for `pesign` and `gpg` signing and
+  # verification to "just work", regardless of which type of
+  # signing profile we have.
+  sbsetup_signing_profile
+
+  # Resign the EFI artifacts.
+  sign_shim "${shim}"
+  sign_mokm "${mokm}"
+  sign_grub "${grub}"
+
+  # Write the resigned artifacts back to the EFI image.
+  mcopy -i "${EFI_IMAGE}" -ov "${EFI_MOUNT}/EFI/BOOT"/*.efi ::/EFI/BOOT
+  mcopy -i "${EFI_IMAGE}" -ov "${SBKEYS}"/db.{crt,cer} ::/EFI/BOOT
+  popd >/dev/null
+
+  # Write the EFI image back to the OS image.
+  check_image_size "${EFI_IMAGE}" "${partsize["EFI-A"]}"
+  dd if="${EFI_IMAGE}" of="${OS_IMAGE}" \
+    conv=notrunc bs=1M seek="${partoff["EFI-A"]}"
+
+  # Resign the kernel and write to boot image.
+  sign_vmlinuz "${BOOT_MOUNT}/vmlinuz"
+  cat <<EOF | debugfs -w -f - "${BOOT_IMAGE}"
+rm vmlinuz
+write ${BOOT_MOUNT}/vmlinuz vmlinuz
+ea_set vmlinuz security.selinux system_u:object_r:os_t:s0
+EOF
+
+  # Generate a new HMAC for the kernel after signing and write to boot image.
+  generate_hmac "${BOOT_MOUNT}/vmlinuz"
+  cat <<EOF | debugfs -w -f - "${BOOT_IMAGE}"
+rm .vmlinuz.hmac
+write ${BOOT_MOUNT}/.vmlinuz.hmac .vmlinuz.hmac
+ea_set .vmlinuz.hmac security.selinux system_u:object_r:os_t:s0
+EOF
+
+fi
+
+###############################################################################
+# Section 6: update boot partition
+
+GRUB_CONFIG="${BOOT_MOUNT}/grub/grub.cfg"
+[[ -s "${GRUB_CONFIG}" ]]
```

Comment:
Do we need this check? It seems a bit lonely - we aren't doing similar checks for other files we expect.

Reply:
We can drop.